### ISSUE TITLE
[Not an Issue] Website version of Lexica

### ISSUE THREAD

**Issue by MattRCole** [lexica/lexica#270]
I hesitate to post this here because issues are supposed to be issues, not discussions. But not finding any references to a message board/discord/etc. I've decided to go ahead and make an `issue`.

[Here](https://mattrcole.github.io/?b=oasrleupaietvbndlhsoznain&l=en_US&t=180&s=l&m=4&mv=20007&v=20014)'s a link to a playable web version of lexica. I've built this over the course of a month and the source code for it, along with a list of its limitations, can be found [here](https://github.com/mattrcole/web-lexica).

I'm posting here to

1. See if there's interest in a complete web version of lexica (at least the multiplayer parts) with support for multiple languages
2. Make sure I'm not stepping on any toes by currently calling it `web-lexica`, and to generally get a feel for any objections (if any) that the current project managers for `Lexica` may have to anything `web-lexica` as it stands today.
3. Because I thought people might be interested in seeing this.

The website can parse the query string in the url of a lexica multiplayer invite link and make a game out of it. All of the logic used in the website is my own, but tries to mirror the behavior of Lexica. Dictionaries are pulled from the `Lexica`'s github repository (but this can be changed).

The purpose and end goal of the project is purely so that I can play multiplayer Lexica games with my friends and family who don't have Android phones.

Give it a look, let me know what you think!

**Comment by pserwylo** [comment]
Oh wow, this is absolutely amazing, great work @MattRCole! The code is very well written too, enjoyable to read (I'm a big user of TypeScript + React in my day job).

Please feel free to do as you wish with this project, but how would you feel about:
* Moving the project under the https://github.com/lexica organisation (of course with you being a developer in that organisation).
* Using this as the official way to let people play without Android by linking to this web version by adding an appropriate link [to this page](https://lexica.github.io/m/?b=oasrleupaietvbndlhsoznain&l=en_US&t=180&s=l&m=4&mv=20007&v=20014#).

I'd be more than happy to help setup a few extra pieces of infrastructure to make this work as smoothy as possible, e.g.
* GitHub Actions to build the web version and make it available via GitHub pages each time `master` is updated (or tagged).
* Online hosting of dictionaries and other information (such as the [letter scores](https://github.com/MattRCole/web-lexica/blob/master/src/game/scores.json)) based on the source information in the https://github.com/lexica/lexica repository. This could also be setup using GitHub Actions to automatically publish updated dictionaries and other metadata when new versions of Lexica are released. I think it is awesome how you link directly to the dictionaries in the Lexica repository, but it wouldn't be that much more effort to have them published in an appropriate directory served via GitHub Pages, and also some additional metadata such as lists of available languages, etc.
* Just generally ongoing support, encouragement, and anything else to help have this as a first class citizen in the Lexica ecosystem.

I mean, long term, there is nothing specifically Android specific about Lexica, so TypeScript + React is actually not such a terrible technology choice for the official game. Perhaps as more work gets put into this version, it could effectively become the normal Android version when it nears/hits parity (and thus those with the means could also help with publishing an iOS version).

Thoughts?

**Comment by MattRCole** [comment]
Yes to all of that! It sounds awesome! I'd love the project to become "official" and I'd love to keep on working with it.

I feel slightly embarrassed at the state of the code. There are a lot of parts of the board logic that need to be broken out in order to support other languages, not to mention a few rather horrendous things I've done to force the game hook to stop re-rendering every frame, but I'm confident that the project can be turned into something much more robust.

The offer of help with GitHub Actions and the hosting of the scoring/dictionaries would be amazing.

I've always been interested in being more active in the open source community, and I think this is a great way to do that. Let me know what I need to do to get the repo transferred over to the `Lexica` organization.

**Comment by pserwylo** [comment]
Awesome, glad to have you onboard. I've created a @lexica/developer team which has write access to the lexica/lexica and lexica/lexica.github.io repositories. When we add lexica/web-lexica I'll add it as a developer on that repository too. I can add you as an administrator on that repository also.

To transfer, you can follow the instructions here: https://docs.github.com/en/github/administering-a-repository/managing-repository-settings/transferring-a-repository#transferring-a-repository-owned-by-your-user-account.

As for what next, happy to take your thoughts on the best way to approach it.

**Comment by MattRCole** [comment]
Okay, I've gotten it transferred over after diving way too deep into the different types of public licenses.

Here are my thoughts on next steps:

- Work to make the web app fully compatible with any multiplayer link
- Decide on color schemes for `web-lexica`, either following `lexica`'s theming or sticking with Solarized (I don't have much of an opinion here)
- Add tests or something similar
  - For the pure logic in the app, typical unit tests would work
  - [Storybook](https://storybook.js.org) stories for the visual components. I personally haven't seen a good way of doing automatic testing of the visual components of an app, so I'm definitely up for any suggestions here.
  - Custom React hooks. I'm not sure how to do testing around hooks and if you have any ideas here, that would be great
- Pausing. How do we want to handle pausing in the web app? Do we handle it at all?

After those things are done and the codebase is in a bit of a more manageable state, then I feel like it would be a good time to start in on implementing more of the logic contained in the original `lexica`.

Some other things that could be done in the future are:
- service workers to keep a responsive UI while the heavy computing is being done in the app
- make it into a Progressive Web Application for offline play. 

Any suggestions? I'm excited to start in on any of this.

**Comment by pserwylo** [comment]
Sounds great, I will comment more fully when I get back from a walk, but thanks for transfering, appreciate the GPL license, and your thoughts above sound good. I've given you full admin writes on the web-lexica repo, so feel free to make full use of GitHub Projects or Issues or any other mechnaism for tracking some of these ideas.

**Comment by pserwylo** [comment]
Toying with some API ideas in #271:
* https://pserwylo.github.io/lexica/api/v1/metadata.json
* https://pserwylo.github.io/lexica/api/v1/language/en_US/metadata.json
* https://pserwylo.github.io/lexica/api/v1/language/en_US/dictionary.json
* https://pserwylo.github.io/lexica/api/v1/language/en_US/dictionary.txt
* https://pserwylo.github.io/lexica/api/v1/language/en_US/probabilities.json
* https://pserwylo.github.io/lexica/api/v1/language/en_US/probabilities.txt

After a bit more playing, will merge into lexica.github.io instead of using my personal fork.

Anything else you think may be useful? Or should the URL format change to be more readable/helpful?

**Comment by MattRCole** [comment]
This looks great! The only thing I think is missing would be the scoring for `Letter Score` games

**Comment by pserwylo** [comment]
Ah yes of course, thanks for the reminder.

Given the amount of info (relatively small), I'll add it to the language specific `metadata.json` file. As I'm writing this, I think I will rename https://pserwylo.github.io/lexica/api/v1/metadata.json to `languages.json` and replace `metadata.json` at the top level with an array (or map) which includes all the language specific `metadata.json` files so they can be fetched in one hit for those who want (again, will be pretty small and compress well).

**Comment by pserwylo** [comment]
Alright, here we go: https://lexica.github.io/lexica/

![image](https://user-images.githubusercontent.com/248565/132212547-e10a5118-7315-4b90-860d-b080b3ae53ea.png)

Hopefully pretty close to its final form, but always happy to take on feedback. I'll clean up the code I used to generate it and hook it up to GitHub Actions later in the week.

**Comment by MattRCole** [comment]
The organization of the endpoints makes sense to me. This will be invaluable for including all of the other languages, and, when we get there, for board generation too.

**Comment by pserwylo** [comment]
Okay, API is up now, and built using GitHub Actions (automatically updated each time `master` is updated in this repo).

Files are at https://lexica.github.io/lexica/. Includes the letter points in the language-specific `metadata.json` files as well as the top level `metadata.json` (this file just contains the combination of all language specific versions).

**Comment by MattRCole** [comment]
That looks excellent and is now being used by the web app. In theory, it should be ready to go for any language supported by Lexica (I'm sure there are errors, but I figure the best way to find them is to load test it)

What base url would you like the web app to live at? Were you thinking of having it use `https://lexica.github.io/lexica/` base path?

**Comment by pserwylo** [comment]
Yes, `lexica.github.io/web-lexica` would be the URL, and also an amendment to https://github.com/lexica/lexica.github.io/blob/master/m/index.html which links people to the web version.

**Comment by GrimPixel** [comment]
Website version will be cool. Polyglot Club is considering integrating games for learning languages on its website. It would be an opportunity if the web version is available.
